File: src/com/company/tree/iterativeTraversal/PostOrder.java
package com.company.tree.iterativeTraversal;

import com.company.tree.TreeNode;

import java.util.*;

public class PostOrder {

    public static void main(String[] args) {
        LinkedList<Integer> list = new LinkedList<>();
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);

        list = postorderTraversal(root, (LinkedList<Integer>) list);

        for (int i = 0; i < list.size(); i++) {
            System.out.print(list.get(i));
            System.out.print(" ");
        }

    }


    public static LinkedList<Integer> postorderTraversal(TreeNode root, LinkedList<Integer> list) {
        LinkedList<Integer> result = new LinkedList<>();
        Deque<TreeNode> stack = new ArrayDeque<>();
        TreeNode curr = root;

        while(!stack.isEmpty() || curr != null) {

            if(curr != null) {
                stack.push(curr);
                result.addFirst(curr.val);  // Reverse the process of preorder
                curr = curr.right;             // Reverse the process of preorder
            } else {
                TreeNode node = stack.pop();
                curr = node.left;           // Reverse the process of preorder
            }

        }
        return result;

      }

    public static LinkedList<Integer> postOrderRecursive(TreeNode root,LinkedList<Integer> linkedList){
        if (root==null){
            return linkedList;
        }
        postOrderRecursive(root.left,linkedList);
        postOrderRecursive(root.right,linkedList);
        linkedList.add(root.val);
        return linkedList;
    }

}

File: src/com/company/tree/iterativeTraversal/PostOrder.kt
package com.company.tree.iterativeTraversal

import com.company.tree.TreeNode
import java.util.*

//Left, Right, Root
fun postorderTraversal(root: TreeNode?): List<Int?> {
    var root = root
    val list = mutableListOf<Int>()
    val stack = Stack<TreeNode>()

    while (stack.isNotEmpty() || root != null) {
        if (root != null) {
            stack.add(root)
            root = root.left
        } else {
            var peekNode = stack.peek().right
            if (peekNode == null) {
                peekNode = stack.pop()
                list.add(peekNode.`val`)
                while (stack.isNotEmpty() && peekNode === stack.peek().right) {
                    peekNode = stack.pop()
                    list.add(peekNode.`val`)
                }
            } else {
                root = peekNode
            }
        }
    }
    return list
}

fun postOrderRecursive(root: TreeNode?, linkedList: MutableList<Int>): MutableList<Int> {
    root ?: return linkedList
    postOrderRecursive(root.left, linkedList)
    postOrderRecursive(root.right, linkedList)
    linkedList.add(root.`val`)
    return linkedList
}

fun main() {
    val root = TreeNode(1)
    val left = TreeNode(2)
    val right = TreeNode(3)
    val leftLeft = TreeNode(4)
    val leftRight = TreeNode(5)

    root.left = left
    root.right = right
    root.left!!.left = leftLeft
    root.left!!.right = leftRight

    val list1 = postorderTraversal(root)
    for (i in list1) {
        print(i)
        print(" ")
    }
}